java/src/test/java/org/rocksdb/FlinkCompactionFilterTest.java

```diff
@@ -0,0 +1,345 @@
+            }
+
+            @Override
+            void updateValueWithTimestamp(RocksDB db) throws RocksDBException {
+                db.merge(columnFamilyHandle, getASCII(key), listExpired(3));
+                db.merge(columnFamilyHandle, getASCII(key), mixedList(2, 3));
+                db.merge(columnFamilyHandle, getASCII(key), listUnexpired(4));
+            }
+
+            @Override
+            byte[] unexpiredValue() {
+                return mixedList(5, 7);
+            }
+
+            byte[] mergeBytes(byte[] ... bytes) {
+                int length = 0;
+                for (byte[] a : bytes) {
+                    length += a.length;
+                }
+                ByteBuffer buffer = ByteBuffer.allocate(length);
+                for (byte[] a : bytes) {
+                    buffer.put(a);
+                }
+                return buffer.array();
+            }
+
+            @Override
+            byte[] expiredValue() {
+                return listUnexpired(7);
+            }
+
+            private byte[] mixedList(int numberOfExpiredElements, int numberOfUnexpiredElements) {
+                assert numberOfExpiredElements > 0;
+                assert numberOfUnexpiredElements > 0;
+                return mergeBytes(
+                        listExpired(numberOfExpiredElements),
+                        new byte[] {DELIMITER},
+                        listUnexpired(numberOfUnexpiredElements));
+            }
+
+            private byte[] listExpired(int numberOfElements) {
+                return list(numberOfElements, currentTime);
+            }
+
+            private byte[] listUnexpired(int numberOfElements) {
+                return list(numberOfElements, currentTime + TTL);
+            }
+
+            private byte[] list(int numberOfElements, long timestamp) {
+                ByteBuffer buffer = getByteBufferForList(numberOfElements);
+                for (int i = 0; i < numberOfElements; i++) {
+                    appendValueWithTimestamp(buffer, userValue, timestamp);
+                    if (i < numberOfElements - 1) {
+                        buffer.put(DELIMITER);
+                    }
+                }
+                return buffer.array();
+            }
+
+            private ByteBuffer getByteBufferForList(int numberOfElements) {
+                int length = ((LONG_LENGTH + INT_LENGTH + userValue.length() + 1) * numberOfElements) - 1;
+                return ByteBuffer.allocate(length);
+            }
+
+            private static class ListElementIter extends FlinkCompactionFilter.AbstractListElementIter {
+                @Override
+                public int nextOffset(int currentOffset) {
+                    int elemLen = ByteBuffer
+                            .wrap(list, currentOffset, list.length - currentOffset)
+                            .getInt(8);
+                    return currentOffset + 13 + elemLen;
```

Comment:
Does `13` means the length of key ? If so,  why `nextOffset` have to include the length of key, I think here should be `currentOffset + elemLen`

Reply:
the full element consists of:
- long timestamp (8 bytes)
- int length of string (4 bytes), see `appendValueWithTimestamp`
- elemLen
- delimiter between elements (1 byte), except last one
13 + elemLen = 8 + 4 + elemLen + 1